src/ddsrt/include/dds/ddsrt/avl.h

```diff
@@ -252,48 +331,440 @@ typedef struct ddsrt_avl_citer {
 
 /* Not maintaining # nodes */
 
+/**
+ * @brief Initialize the @ref ddsrt_avl_treedef
+ * 
+ * Different types of trees are supported. This function defines which type to use.
+ * Given a node defined as:
+ * @code{.c}
+ * typedef struct num_s{ // user node
+ *   ddsrt_avl_node_t node;
+ *   uint64_t val; // user data
+ * }num_t;
+ * @endcode
+ * the offsets are obtained as:
+ * @code{.c}
+ * size_t avlnodeoffset = offsetof(num_t, node);
+ * size_t keyoffset = offsetof(num_t, val);
+ * @endcode
+ * 
+ * Regarding the flags:
+ * - The flag @ref DDSRT_AVL_TREEDEF_FLAG_INDKEY affects functions like @ref ddsrt_avl_compare_t and @ref ddsrt_avl_lookup.
+ *   These functions expect their 'key' argument to be a pointer, so by default you need to pass a reference to the key.
+ *   Alternatively, by using @ref DDSRT_AVL_TREEDEF_FLAG_INDKEY you indicate that the key is already a pointer, allowing you to pass the key directly.
+ *   As an example in the case of a C string, using the INDKEY flag means you can provide the key directly (char*), rather than a reference (char**).
```

Comment:
"It might also be useful to make explicit that setting keyoffset to 0 means you have to pass a "dummy" node with only the key filled in when calling lookup"

I'm not sure if I'm following. Doesn't keyoffset == 0 mean that the key is the first element in the node?

Reply:
Well, yes — and no. There is the model of how this is expected to be used most of the time and there it does indeed mean it is the first element in the node. (Which is fine, the `avl_node_t` doesn't have to be the first; and indeed you can have multiple `avl_node_t`s in one node and maintain multiple trees.)

But what keyoffset = 0 really means is that the compare function is passed the address of the node. Nothing ever interprets the data at address `(char*)node+keyoffset` (*) except the compare function that the user provides. Therefore all you really must do is make sure that this function accepts a pointer to whatever resides at `keyoffset`, and that a pointer of the same type is passed to `lookup` so that the compare function is an instance of the `Ptr t -> Ptr t -> Ordering` type.

The two obvious possibilities are: the actual key value and `t` is the type of the key; and the entire node and `t` is the type of the node.

(*) Arguably with the exception of `INDKEY` which interprets the contents at that address as `Ptr (Ptr t)` and dereferences it once before passing it to the compare function.